topics/search.md

```diff
@@ -0,0 +1,167 @@
+@color:{hello world | hello universe}
+```
+
+### Numeric Index
+
+Numeric indexes allow for filtering queries to only return values that are in between a given start and end value.
+
+- Both inclusive and exclusive queries are supported.
+- For open ended queries, `+inf`, `-inf` can be used to express start and end ranges.
+
+As an example, the following query will return books published between 2021 and 2024 (Both inclusive).
+The equivalent mathematical expression is `2021 <= year <= 2024`.
+
+```
+"@year:[2021 2024]"
+```
+
+While The following query will return books published between 2021 (exclusive) and 2024 (inclusive).
+The equivalent mathematical expression is `2021 < year <= 2024`:
+
+```
+"@year:[(2021 2024]"
+```
+
+The following query will return books published before 2024 (inclusive). The equivalent mathematical expression is year `<= 2024`:
+
+```
+@year:[(-inf 2024]
+```
+
+The following query will return books published after 2015 (exclusive). The equivalent mathematical expression is year `>= 2015`:
+
+```
+@year:[2015 +inf]
+```
+
+### Query planner
+
+There are two primary approaches to hybrid queries:
+
+- **Pre-filtering:** Begin by filtering the dataset and then perform an exact similarity search. This works well when the filtered result set is small but can be costly with larger datasets.
+- **Post-filtering:** Perform the similarity search first, then filter the results. This is suitable when the filter-qualified result set is large but may lead to empty or lower than expected amount of results.
+
+Valkey-Search uses a **hybrid approach** with a query planner that selects the most efficient query execution path between:
+
+- **Pre-filtering**
+- **Inline-filtering:** Filters results during the similarity search process.
+
+## Monitoring
+
+To check the server's overall search metrics, you can use the `INFO SEARCH` or `INFO MODULES` commands.
+
+The following metrics are added to the `INFO` command's output:
+
+- **search_used_memory_human** (Integer) A human-friendly readable version of the `search_used_memory_bytes` metric
+- **search_used_memory_bytes** (Integer) The total bytes of memory that all indexes occupy
+- **search_number_of_indexes** (Integer) Index schema total count
+- **search_number_of_attributes** (Integer) Total count of attributes for all indexes
+- **search_total_indexed_hash_keys** (Integer) Total count of HASH keys for all indexes
+- **search_background_indexing_status** (String) The status of the indexing process. `NO_ACTIVITY` indicates idle indexing.
+- **search_failure_requests_count** (Integer) A count of all failed requests, including syntax errors.
+- **search_successful_requests_count** (Integer) A count of all successful requests
+- **search_hnsw_create_exceptions_count** (Integer) Count of HNSW creation exceptions.
+- **search_hnsw_search_exceptions_count** (Integer) Count of HNSW search exceptions
+- **search_hnsw_remove_exceptions_count** (Integer) Count of HNSW removal exceptions.
```

Comment:
What is the definition of an exception? We almost already refer to commands as errors everywhere else in the code base.